### PR TITLE
Add generate csr function

### PR DESCRIPTION
**Pull request by wattli**

### CODE REVIEW

File: cmd/generate_csr/main.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2017 Istio Authors
```
Comment (lookuptable) [inline cmd/generate_csr/main.go]: Just curious: who is going to call this executable?

File: certmanager/generate_cert.go [hunk 2]
```diff
@@ -142,6 +168,22 @@ func genSerialNum() *big.Int {
 	return serialNum
 }
 
+// genCertTemplate generates a certificateRequest template with the given options.
```
Comment (lookuptable) [inline certmanager/generate_cert.go]: s/genCertTemplate/genCsrTemplate

File: certmanager/generate_cert.go [hunk 3]
```diff
@@ -83,6 +83,32 @@ const (
 // See http://www.alvestrand.no/objectid/2.5.29.17.html.
 var oidSubjectAltName = asn1.ObjectIdentifier{2, 5, 29, 17}
 
+// GenCSR generates a X.509 certificate sign request with the given options.
```
Comment (lookuptable) [inline certmanager/generate_cert.go]: This method returns both a CSR and the private key. Can you update the comment?

I notice we have a similar issue on line 112. I'll fix that later

File: certmanager/generate_cert.go [hunk 4]
```diff
@@ -83,6 +83,32 @@ const (
...
+func GenCsr(options CertOptions) ([]byte, []byte) {
```
Comment (lookuptable) [inline certmanager/generate_cert.go]: s/GenCsr/GenCSR

File: certmanager/generate_cert.go [hunk 5]
```diff
@@ -83,6 +83,32 @@ const (
...
+	// Generates a CSR
+	// The public key will be bound to the certficate generated below. The
+	// private key will be used to sign this certificate in the self-signed
+	// case, otherwise the certificate is signed by the signer private key
+	// as specified in the CertOptions.
```
Comment (lookuptable) [inline certmanager/generate_cert.go]: Some of the comments are not appropriate here - this method should only generate a CSR and not worry about who is going to sign it.

File: certmanager/generate_cert.go [hunk 6]
```diff
@@ -83,6 +83,32 @@ const (
...
+		glog.Fatalf("RSA key generation failed with error %s.", err)
+	}
+	template := genCsrTemplate(options)
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, &template, crypto.PrivateKey(priv))
+	if err != nil {
+		glog.Fatalf("Could not create certificate (err = %s).", err)
+	}
+
+	// Returns the certificate that carries the RSA public key as well as
+	// the corresponding private key.
```
Comment (sarvaniv) [inline certmanager/generate_cert.go]: The comment says the certificate carries RSA public key as well as private key. This does not sound right to me. The CSR should contain a digital signature signed with the private key. Could you clarify?

File: cmd/generate_csr/main.go (hunk 1 above)
Comment (wattli) [inline cmd/generate_csr/main.go]: The node agent that lives in node, see https://docs.google.com/document/d/1RZnBfj8I5xs8Yi_BPEKBRp0K3UnIJYTDg_31rfTt4j8/edit#heading=h.p0zc37bjhmkg

File: certmanager/generate_cert.go (hunk 3 above)
Comment (wattli) [inline certmanager/generate_cert.go]: Both are updated since i am already here.

File: certmanager/generate_cert.go (hunk 5 above)
Comment (wattli) [inline certmanager/generate_cert.go]: Updated.

File: certmanager/generate_cert.go (hunk 2 above)
Comment (wattli) [inline certmanager/generate_cert.go]: Done

File: certmanager/generate_cert.go (hunk 3 above)
Comment (sarvaniv) [inline certmanager/generate_cert.go]: nice :-)

File: certmanager/generate_cert.go (hunk 6 above)
Comment (wattli) [inline certmanager/generate_cert.go]: Reworded. It actually means returns a CSR and a private key

File: certmanager/generate_cert.go (hunk 4 above)
Comment (sarvaniv) [inline certmanager/generate_cert.go]: As discussed, lets not replicate code. Some refactoring will help keep common code in a function.

Comment (wattli) [inline certmanager/generate_cert.go]: done

File: certmanager/generate_cert_test.go [hunk 7]
```diff
@@ -36,6 +36,36 @@ type VerifyFields struct {
 
 var now = time.Now().Round(time.Second).UTC()
 
+func TestGenCsr(t *testing.T) {
```
Comment (sarvaniv) [inline certmanager/generate_cert_test.go]: Do we need tests for negative case? For example, invalid csr options?
